nan.h

```diff
@@ -136,15 +136,77 @@ namespace Nan {
 
 //=== RegistrationFunction =====================================================
 
-#if NODE_MODULE_VERSION < IOJS_3_0_MODULE_VERSION
-  typedef v8::Handle<v8::Object> ADDON_REGISTER_FUNCTION_ARGS_TYPE;
+typedef void (*addon_reg_func_t)(v8::Local<v8::Object> target);
+
+#if NODE_MODULE_VERSION < NODE_0_10_MODULE_VERSION
+namespace imp {
+  template<addon_reg_func_t F>
+  static inline void addon_reg_func(v8::Handle<v8::Object> target) {
+    v8::HandleScope scope;
+    F(v8::Local<v8::Object>::New(target));
+  }
+}
+
+# define NAN_MODULE(modname, regfunc)                                          \
+  extern "C" {                                                                 \
+    NODE_MODULE_EXPORT node::node_module_struct modname ## _module =           \
+    {                                                                          \
+      NODE_STANDARD_MODULE_STUFF,                                              \
+      Nan::imp::addon_reg_func<regfunc>,                                       \
+      NODE_STRINGIFY(modname)                                                  \
+    };                                                                         \
+  }
+
+#elif NODE_MODULE_VERSION < NODE_0_12_MODULE_VERSION
+namespace imp {
+  template<addon_reg_func_t F>
+  static inline void addon_reg_func(v8::Handle<v8::Object> target) {
+    v8::HandleScope scope;
+    F(v8::Local<v8::Object>::New(target));
+  }
+}
+
+# define NAN_MODULE(modname, regfunc)                                          \
+  extern "C" {                                                                 \
+    NODE_MODULE_EXPORT node::node_module_struct modname ## _module =           \
+    {                                                                          \
+      NODE_STANDARD_MODULE_STUFF,                                              \
+      (node::addon_register_func) Nan::imp::addon_reg_func<regfunc>,           \
+      NODE_STRINGIFY(modname)                                                  \
+    };                                                                         \
+  }
+#elif NODE_MODULE_VERSION < IOJS_3_0_MODULE_VERSION
+namespace imp {
+  template<addon_reg_func_t F>
+  static inline void addon_reg_func(v8::Handle<v8::Object> target) {
+    v8::Isolate *isolate = v8::Isolate::GetCurrent();
+    v8::HandleScope scope(isolate);
+    F(v8::Local<v8::Object>::New(isolate, target));
+  }
+}
+
+# define NAN_MODULE(modname, regfunc)                                          \
+  extern "C" {                                                                 \
```

Comment:
Do we have to duplicate this macro? Isn't it enough to call it with `Nan::imp::addon_reg_func<regfunc>` as an argument?